cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,243 @@
+	// Only used in test.
+	failpointCh chan error
+}
+
+func (k *kafkaProducer) AsyncSendMessage(
+	ctx context.Context, topic string,
+	partition int32, message *codec.MQMessage,
+) error {
+	// We have to hold the lock to prevent write to closed producer.
+	// Close may be blocked for a long time.
+	k.closedMu.RLock()
+	defer k.closedMu.RUnlock()
+
+	// If the producer is closed, we should skip the message.
+	if k.closed {
+		// We return the context error directly rather than other error.
+		// Because if producer already closed, it means the context maybe is canceling or canceled.
+		// So we shouldn't return other errors to the caller to mess up the shutdown process.
+		return ctx.Err()
+	}
+	failpoint.Inject("KafkaSinkAsyncSendError", func() {
+		// simulate sending message to input channel successfully but flushing
+		// message to Kafka meets error
+		log.Info("failpoint error injected", zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		k.failpointCh <- errors.New("kafka sink injected error")
+		failpoint.Return(nil)
+	})
+
+	failpoint.Inject("SinkFlushDMLPanic", func() {
+		time.Sleep(time.Second)
+		log.Panic("SinkFlushDMLPanic",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+	})
+
+	msg := &sarama.ProducerMessage{
+		Topic:     topic,
+		Partition: partition,
+		Key:       sarama.StringEncoder(message.Key),
+		Value:     sarama.ByteEncoder(message.Value),
+		Metadata:  messageMetaData{callback: message.Callback},
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.asyncProducer.Input() <- msg:
+	}
+	return nil
+}
+
+func (k *kafkaProducer) Close() {
+	// We have to hold the lock to prevent write to closed producer.
+	k.closedMu.Lock()
+	defer k.closedMu.Unlock()
+	// If the producer was already closed, we should skip the close operation.
+	if k.closed {
+		// We need to guard against double closed the clients,
+		// which could lead to panic.
+		log.Warn("kafka producer already closed",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID),
+			zap.Any("role", k.role))
+		return
+	}
+	close(k.failpointCh)
+	// Notify the run loop to exit.
+	close(k.closedChan)
+	k.closed = true
+	// We need to close it asynchronously.
+	// Otherwise, we might get stuck with it in an unhealthy state of kafka.
+	go func() {
+		// `client` is mainly used by `asyncProducer` to fetch metadata and other related
+		// operations. When we close the `kafkaSaramaProducer`, TiCDC no need to make sure
+		// that buffered messages flushed.
+		// Consider the situation that the broker does not respond, If the client is not
+		// closed, `asyncProducer.Close()` would waste a mount of time to try flush all messages.
+		// To prevent the scenario mentioned above, close client first.
+		start := time.Now()
+		if err := k.client.Close(); err != nil {
```

Comment:
- Will `client.Close()` take long?
- Since the closing takes place asynchronously in a goroutine, how do we avoid a zombie problem? After `Close()` has returned, the developer of the caller may want to make sure that no new data will be written to the downstream. How can this be implemented? Is this a risk?

Reply:
1. possible(maybe about 1-2 mins)
2. Yes, it's risk. But it's acceptable.
 - If the kafka cluster is running well, it will be shut down as soon as possible. Also, we cancel all tablepiples before shutdown, so it's safe.
 - If there is a problem with the kafka cluster, it will shut down the client first, which means no more data will be sent because the connection to the broker is dropped. Also, all our tablepiples will be closed before the shutdown, so it is safe.
 - For mq, duplicate data is acceptable.
 - There is a risk of goroutine leakage, but it is acceptable and our main goal is not to get stuck with the tick.

Reply:
Added safety note. Please checkout [1b5cd1f](https://github.com/pingcap/tiflow/pull/6412/commits/1b5cd1ff0d6819c225e16660b319ef7e91a14441)

Reply:
> 1. possible(maybe about 1-2 mins)
> 2. Yes, it's risk. But it's acceptable.
> 
> * If the kafka cluster is running well, it will be shut down as soon as possible. Also, we cancel all tablepiples before shutdown, so it's safe.
> * If there is a problem with the kafka cluster, it will shut down the client first, which means no more data will be sent because the connection to the broker is dropped. Also, all our tablepiples will be closed before the shutdown, so it is safe.
> * For mq, duplicate data is acceptable.
> * There is a risk of goroutine leakage, but it is acceptable and our main goal is not to get stuck with the tick.

- Can we guarantee that no data will be written to the downstream if the table pipelines have been canceled but the Kafka producer is not yet closed?
- Duplication is acceptable only in the sense that the table is never being replicated by more than one TiCDC node. My concern is that before the client on one TiCDC node is completely closed, the table is scheduled to another TiCDC node.
- The interactions with table pipelines you describe seem like very complex coupling. What is a mitigation strategy? Is there an integration test case available?

Reply:
> * Can we guarantee that no data will be written to the downstream if the table pipelines have been canceled but the Kafka producer is not yet closed?

I think it is impossible in most cases. 1. if the kafka cluster is normal, we shut down the client first, which will disconnect the connection so no data can be written. 2. if the kafka cluster is not working, the data cannot be written. 3. the kafka cluster suddenly recovers, but we also shut down the client first so no more data will be written.

But the shutdown of the client and the sending of data from the producer are asynchronous. So there is still a possibility that the table that is scheduled to go away is still being written during the process. I'm not quite sure how to fix it more directly. Unless we modify the tick mechanism.



> * The interactions with table pipelines you describe seem like very complex coupling. What is a mitigation strategy? Is there an integration test case available?

We have some tests that inject sink errors that should cover it. But I can't say exactly how it covers these code paths.

Reply:
> So there is still a possibility that the table that is scheduled to go away is still being written during the process. I'm not quite sure how to fix it more directly. Unless we modify the tick mechanism.

I think the probability of this is very low, because the producer always needs to get the leader through the client before sending a message (in case of failure). So when the cluster recovers, it should also have a hard time sending out messages. But in general it is possible, for example, that the producer finds a leader that has just recovered and is available when the client's shutdown operation has just been executed and the lock has not yet been obtained. In general it is difficult to find such a broker to send messages.

Reply:
Offline discussion:

Maybe we need to fix this issue at sarama client. We can drop the retry buffers directly.

Reply:
Tracking in https://github.com/pingcap/tiflow/issues/6472

We'll discuss this issue on Wed.